backend/entityservice/database/insertions.py

```diff
@@ -261,3 +259,28 @@ def get_created_runs_and_queue(db, project_id):
     if res is None:
         res = []
     return res
+
+
+def is_dataprovider_allowed_to_upload_and_lock(db, dp_id):
+    """
+    This method returns true if the dataprovider is allowed to upload her clks.
+    A data provider is allowed to upload clks if they have not yet been uploaded or if the upload is already in progress.
+    This method will lock the resource by setting the upload state to `in_progress` and returning `true`.
+    Note that the upload state can be `error`, in which case we are allowing the dataprovider to re-try uploading
+    her clks not to block a project if a failure occurred.
+    """
+    logger.debug("Setting dataprovider {} upload state to `in_progress``".format(dp_id))
+    sql_update = """
+        UPDATE dataproviders
+        SET uploaded = 'in_progress'
+        WHERE id = %s and uploaded != 'done' and uploaded != 'in_progress'
+        RETURNING id, uploaded
+        """
+    query_response = query_db(db, sql_update, [dp_id])
+    print(query_response)
+    length = len(query_response)
+    if length < 1:
+        return False
+    elif length > 1:
+        raise ValueError("Houston, we have a problem!!! This dataprovider can upload multiple times her clks.")
```

Comment:
it would be good to write a message to the logger at high severity level as well here.